scipy/optimize/_root_scalar.py

```diff
@@ -0,0 +1,237 @@
+"""
+Unified interfaces to root finding algorithms for real or complex
+scalar functions.
+
+Functions
+---------
+- root : find a root of a scalar function.
+"""
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+from scipy._lib.six import callable
+
+from . import zeros as optzeros
+
+__all__ = ['root_scalar']
+
+
+class MemoizeDer(object):
+    """Decorator that caches the value and derivative(s) of function each
+    time it is called.
+
+    This is a simplistic memoizer that calls and caches a single value
+    of `f(x, *args)`.
+    It assumes that `args` does not change between invocations.
+    It supports the use case of a root-finder where `args` is fixed,
+    `x` changes, and only rarely, if at all, does x assume the same value
+    more than once."""
+    def __init__(self, fun):
+        self.fun = fun
+        self.vals = None
+        self.x = None
+        self.n_calls = 0
+
+    def __call__(self, x, *args):
+        r"""Calculate f or used cached value if available"""
+        # Derivative may be requested before the function itself, always check
+        if self.vals is None or x != self.x:
+            fg = self.fun(x, *args)
+            self.x = x
+            self.n_calls += 1
+            self.vals = fg[:]
+        return self.vals[0]
+
+    def fprime(self, x, *args):
+        r"""Calculate f' or used a cached value if available"""
+        if self.vals is None or x != self.x:
+            self(x, *args)
+        return self.vals[1]
+
+    def fprime2(self, x, *args):
+        r"""Calculate f'' or used a cached value if available"""
+        if self.vals is None or x != self.x:
+            self(x, *args)
+        return self.vals[2]
+
+    def ncalls(self):
+        return self.n_calls
+
+
+def root_scalar(f, args=(), method=None, bracket=None,
+                fprime=None, fprime2=None,
+                x0=None, x1=None,
+                xtol=None, rtol=None, maxiter=None,
+                options=None):
+    """
+    Find a root of a scalar function.
+
+    Parameters
+    ----------
+    f : callable
```

Comment:
perhaps `fun`? Since this #8890 is supposed to address #8884 "inconsistent names of function arguments" then shouldn't `root_scalar` use the same keyword arguments as `minimize`, `minimize_scalar`, and `root` which all use `fun`? Also I think in general (and there are always exceptions) pep-8 discourages the use of single letter variables names

okay, I see, all of the scalar roots except newton are currently using `f` so maybe that is the prevailing convention. I'm fine either way.